src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsTest.java

```diff
@@ -0,0 +1,268 @@
+                        .withStatus(404)
+                        .withStatusMessage("404 Not Found")
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBody("{\"message\": \"File not found\"}")));
+
+        githubApi.stubFor(
+            post(urlEqualTo("/app/installations/654321/access_tokens"))
+                .inScenario(scenarioName)
+                .whenScenarioStateIs("3")
+                .willSetStateTo("4")
+                .withRequestBody(
+                    equalToJson(
+                        "{\"permissions\":{\"pull_requests\":\"write\",\"metadata\":\"read\",\"checks\":\"write\",\"contents\":\"read\"}}", true, false))
+                .willReturn(
+                    aResponse()
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBodyFile("../AppCredentials/files/body-githubapp-create-installation-accesstokens.json")));
+
+    }
+
+    @Test
+    public void testAgentRefresh() throws Exception {
+        long notStaleSeconds = GitHubAppCredentials.AppInstallationToken.NOT_STALE_MINIMUM_SECONDS;
+        try {
+            appCredentials.setApiUri(githubApi.baseUrl());
+
+            // We want to demonstrate successful caching without waiting for a the default 1 minute
+            // Must set this to a large enough number to avoid flaky test
+            GitHubAppCredentials.AppInstallationToken.NOT_STALE_MINIMUM_SECONDS = 15;
+
+            final String jenkinsfile = String.join(
+                "\n",
+                "// run checkout several times",
+                "node ('my-agent') {",
+                // First Checkout on agent should use cached token passed via remoting
+                "    checkout scm",
+                // Multiple checkouts in quick succession should cached token
+                "    checkout scm",
+                "    sleep " + (GitHubAppCredentials.AppInstallationToken.NOT_STALE_MINIMUM_SECONDS + 2),
+                // Checkout after token expires refreshes via remoting - error on controller is not catastrophic
+                "    checkout scm",
+                // Checkout after error will refresh again on controller
+                "    checkout scm",
+                // Multiple checkouts in quick succession should use cached token
+                "    checkout scm",
+                "}");
+
+
+            sampleRepo.init();
+            sampleRepo.write("Jenkinsfile", jenkinsfile);
+            sampleRepo.git("add", "Jenkinsfile");
+            sampleRepo.git("commit", "--message=init");
+
+            WorkflowJob job = r.createProject(WorkflowJob.class, "repo-master");
+            GitStep gitStep = new GitStep(sampleRepo.toString());
+            gitStep.setCredentialsId(myAppCredentialsId);
+            gitStep.setPoll(false);
+            job.setDefinition(new CpsScmFlowDefinition(gitStep.createSCM(), "Jenkinsfile"));
```

Comment:
Seems convoluted. If you are not testing `GitHubSCMSource` anyway, why not just use the usual `CpsFlowDefinition` with the desired script, and replace

```groovy
checkout scm
```

with

```
git url: REPO, credentialsId: 'hard-coded-id'
```

after adding a `StringParameterDefinition` with the repo location?

Reply:
(BTW the `StringParameterDefinition` might seem more complicated than just using a Groovy string literal inserting the repo location into the script from the Java test code, but that shortcut can run into problems with Windows paths involving `\`. Sometimes I have used shelly strings `"… url: $/" + location + "/$, …"` but this is not foolproof either, and is harder to read.)

Reply:
@jglick 
I want it to checkout the Jenkinsfile using the credential to when it starts the pipeline. I suppose I could call `getPassword()` manually once to force things and then do as you suggest.

Reply:
Not important, just wondering if there was a reason to set things up this way.

Reply:
Fixing.